src/meta/meta_ckpt.c

```diff
@@ -503,42 +538,213 @@ __wt_meta_blk_mods_load(WT_SESSION_IMPL *session, const char *config, WT_CKPT *c
+         ckpt_saved++, ckpt_new++)
+        __assert_ckpt_matches(session, ckpt_saved, ckpt_new);
+
+    if (!((ckpt_saved == NULL && ckpt_new == NULL) ||
+          ((ckpt_saved != NULL && ckpt_saved->order == 0) &&
+            (ckpt_new != NULL && ckpt_new->order == 0))))
+        WT_ASSERT(session, false);
+}
+#endif
+
+/*
+ * __meta_ckptlist_allocate_new_ckpt --
+ *     Provided a checkpoint list, allocate a new checkpoint. Either use the last checkpoint in the
+ *     list or the file metadata to initialize this new checkpoint.
+ */
+static int
+__meta_ckptlist_allocate_new_ckpt(
+  WT_SESSION_IMPL *session, WT_CKPT **ckptbasep, size_t *allocated, const char *config)
+{
+    WT_CKPT *ckptbase, *ckpt;
+    WT_CONNECTION_IMPL *conn;
+    size_t slot;
+    uint64_t most_recent;
+
+    ckptbase = *ckptbasep;
+    conn = S2C(session);
+    slot = 0;
+
+    if (ckptbase != NULL)
+        WT_CKPT_FOREACH (ckptbase, ckpt)
+            slot++;
+
+    /* Either we have a configuration or an existing checkpoint to initialize with. */
+    WT_ASSERT(session, config != NULL || slot != 0);
+
+    /*
+     * If we are using an existing checkpoint, we must have the associated metadata, otherwise we
+     * will have to go slow path and read the metadata.
+     */
+    if (config == NULL && ckptbase[slot - 1].block_metadata == NULL)
+        WT_RET(WT_NOTFOUND);
+
+    /*
+     * This isn't clean, but there's necessary cooperation between the schema layer (that maintains
+     * the list of checkpoints), the btree layer (that knows when the root page is written, creating
+     * a new checkpoint), and the block manager (which actually creates the checkpoint). All of that
+     * cooperation is handled in the array of checkpoint structures referenced from the WT_BTREE
+     * structure.
+     *
+     * Allocate a slot for a new value, plus a slot to mark the end.
+     */
+    WT_RET(__wt_realloc_def(session, allocated, slot + 2, &ckptbase));
+    *ckptbasep = ckptbase;
+
+    ckpt = &ckptbase[slot];
+    ckpt->order = (slot == 0) ? 1 : ckptbase[slot - 1].order + 1;
+    __wt_seconds(session, &ckpt->sec);
+    /*
+     * Update time value for most recent checkpoint, not letting it move backwards. It is possible
+     * to race here, so use atomic CAS. This code relies on the fact that anyone we race with will
+     * only increase (never decrease) the most recent checkpoint time value.
+     */
+    for (;;) {
+        WT_ORDERED_READ(most_recent, conn->ckpt_most_recent);
+        if (ckpt->sec <= most_recent ||
+          __wt_atomic_cas64(&conn->ckpt_most_recent, most_recent, ckpt->sec))
+            break;
+    }
+
+    /* Either load block mods from the config, or from the previous checkpoint. */
+    if (config != NULL)
+        WT_RET(__wt_meta_blk_mods_load(session, config, NULL, ckpt, false));
+    else
+        WT_RET(__wt_meta_blk_mods_load(session, NULL, &ckptbase[slot - 1], ckpt, false));
```

Comment:
This conditional could be eliminated and become just one call to that funciton. If `config != NULL` then it doesn't matter if you pass in a ckptbase, right? It is ignored. And sending in `config` will either be NULL or not.

Reply:
Yes, in an earlier change I had it like that, and I preferred it, but it was not very readable. Also I am being overtly cautious with this change and asserted that either config, or previous checkpoint is passed to the function.

I will make the change, have another look at it and see if you like it.